src/dht_proxy_client.cpp

```diff
@@ -864,67 +864,70 @@ DhtProxyClient::restartListeners()
             {
                 auto settings = std::make_shared<restbed::Settings>();
                 std::chrono::milliseconds timeout(std::numeric_limits<int>::max());
                 settings->set_connection_timeout(timeout); // Avoid the client to close the socket after 5 seconds.
 
                 auto ok = std::make_shared<std::atomic_bool>(true);
                 restbed::Http::async(req,
-                    [this, filter, cb, ok](const std::shared_ptr<restbed::Request>& req,
+                    [this, filter, cb, state](const std::shared_ptr<restbed::Request>& req,
                                            const std::shared_ptr<restbed::Response>& reply) {
                     auto code = reply->get_status_code();
 
                     if (code == 200) {
                         try {
-                            while (restbed::Http::is_open(req)) {
+                            while (restbed::Http::is_open(req) and not state->cancel) {
                                 restbed::Http::fetch("\n", reply);
+                                if (state->cancel)
+                                    break;
                                 std::string body;
                                 reply->get_body(body);
                                 reply->set_body(""); // Reset the body for the next fetch
 
                                 Json::Value json;
                                 std::string err;
                                 Json::CharReaderBuilder rbuilder;
-                                auto* char_data = reinterpret_cast<const char*>(&body[0]);
                                 auto reader = std::unique_ptr<Json::CharReader>(rbuilder.newCharReader());
-                                if (reader->parse(char_data, char_data + body.size(), &json, &err)) {
+                                if (reader->parse(body.data(), body.data() + body.size(), &json, &err)) {
                                     auto value = std::make_shared<Value>(json);
-                                    if ((not filter or filter(*value)) && cb) {
-                                        auto okCb = std::make_shared<std::promise<bool>>();
-                                        auto futureCb = okCb->get_future();
-                                        {
-                                            std::lock_guard<std::mutex> lock(lockCallbacks);
-                                            callbacks_.emplace_back([cb, value, okCb](){
-                                                okCb->set_value(cb({value}, false));
-                                            });
-                                            loopSignal_();
-                                        }
-                                        futureCb.wait();
-                                        if (!futureCb.get()) {
-                                            return;
-                                        }
+                                    if (not filter or filter(*value))  {
```

Comment:
Just need to try that cb is not nullptr.

Reply:
Done